Server post

adamfweidman opened the pull request:
Finished all of Step 4

adamfweidman committed:
Add doPost function and edit doGet so properly returns format on /data URL

adamfweidman committed:
make sure the getData() function is correct and button is onclick

adamfweidman committed:
Merge branch 'Server-JSON' into Server-POST

adamfweidman committed:
Correct typos

bootsieh commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -25,20 +25,40 @@
 public class DataServlet extends HttpServlet {
-  
-  
+  //ArrayList<String> inputs = new ArrayList<String>();
```
forgot to remove this comment

bootsieh commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -25,20 +25,40 @@
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  public void doGet(HttpServletRequest request, HttpServletResponse response) 
+  throws IOException {
```
see my comment below on indenting when a method declaration gets too long

bootsieh commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -25,20 +25,40 @@
-    String json = gson.toJson(messages);
+    /**Gson gson = new Gson();
+    String output = gson.toJson(inputs);
+    response.getWriter().println(output);**/
```
remove commented out code

bootsieh commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -25,20 +25,40 @@
+      String output = userName + ":"  + "\n" + userComment;
+      response.getWriter().println(output);
+    }
+      //response.sendRedirect("/index.html");**/
```
remove commented out code

bootsieh commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -25,20 +25,40 @@
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) 
+  throws IOException {
```
If the line is too long (over 80 chars), then I'd suggest either keeping "throws IOException {" on its own line, but indenting it at least 4 spaces or formatting like this:

public void doPost(HttpServletRequest request,
                              HttpServletResponse response) throws IOException {

It's really up to you, there isn't a firm set of rules on how to break up a method declaration as far as I know, we just don't want the wrapped line to start at the same indent as the start of the line. So whatever looks best to you, go with that and just be consistent.

bootsieh replied:
make sure you resolve this comment before sending for the next review!

bootsieh commented on portfolio/src/main/webapp/index.html:
```diff
@@ -85,18 +85,21 @@ <h1>Adam's Personal Portfolio</h1>
+         <input type="text" name="user-name" required>
+         </br></br>
+
+         //change format for comment box
```
I think you accidentally switched from the HTML specific comments to the '//' one, so this is showing in your UI again now

adamfweidman committed:
Merge branch 'Server-JSON' into Server-POST

bootsieh commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -26,19 +26,41 @@
+    //patches together the name and comment
+    for (int i = 0; i < names.size(); i++) {
+      Gson gson = new Gson();
+      String userName = gson.toJson(names.get(i));
+      String userComment = gson.toJson(comments.get(i));
```
I don't know if two array lists is the best data structure for this scenario. You'll always want to associate names.get(i) and comments.get(i), so it would be good to make that association less fragile (i.e. make them part of the same structure). There are a couple of ways to do this, but I would suggest making some sort of small object to encompass that information.

adamfweidman replied:
Do you suggest making an object  (eg User)that contains, for example, a username and a comment and then I can iterate through those users? Or rather make an arrayList that contains both the username and the comment separated by a special character or something

bootsieh replied:
I think a whole object for a User might be overkill at this point (but good thought!), since we don't really have a firm concept of a 'user' yet, and that brings up a lot of questions about what a User is and if you should track names and combine comments for the same user. I was thinking an even simpler object than that, to just track a single comment.

Saving the string 'userName + ":" + userComment' in a list isn't a bad idea for this CL, however I don't know how sustainable that will be, especially since you're saving the name and comment separately in the entity in your next CL: https://github.com/adamfweidman/myStepPortfolio/pull/7/files.

bootsieh reviewed (commented):
I think it would be good to spend a few minutes improving the UI of the comment boxes. The way it is now has two buttons very close together and without a clear organization:

bootsieh reviewed (changes requested):
I accidentally clicked 'Approve' instead of 'Request changes', so this 'Request changes' applies to my previous comments.

adamfweidman committed:
Implement Bootsie's comments, mostly style

adamfweidman committed:
remove comments

adamfweidman committed:
fixing merge conflicts

bootsieh reviewed (commented):
For future CLs, it would be great if you gave them a bit more of a descriptive name that describes more of what's being changed. It makes it a lot easier for us as reviewers to understand exactly what the purpose of the change is and differentiate it from other changes you've sent out.

Thanks!

adamfweidman committed:
Create User class w name and comment field, implement this change in the servlet

adamfweidman committed:
Fixed the printing format of the comments

adamfweidman commented:
In order to address many of the issues I added a User class which ended up being very simple. This made it much easer for me use json to print the comments to the page. I changed the script to print the comments in a better way.

tristan-jh commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -20,25 +20,33 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  
+  ArrayList<User> users = new ArrayList<User>();
```
prefer using the List interface on the left side of the assignment.
Also! One of my favorite recent(ish) changes to the java language is that you can now skip the type specification on the right side. eg:
List\<User\> = new ArrayList<>();

full details on when this is allowed are here: https://docs.oracle.com/javase/8/docs/technotes/guides/language/type-inference-generic-instance-creation.html

tristan-jh commented on portfolio/src/main/java/com/google/sps/data/User.java:
```diff
@@ -0,0 +1,26 @@
+
+package com.google.sps.data;
+
+public class User {
```
nit: add a comment describing what this class is. Just a sentence or two is plenty for a simple class like this.

bootsieh replied:
I'm not crazy about the name "User", it implies that this is tracking an actual user, which it isn't, just the comment someone made. How about the name "Comment" instead?

bootsieh replied:
Add a class comment describing what this class is for

tristan-jh commented on portfolio/src/main/java/com/google/sps/data/User.java:
```diff
@@ -0,0 +1,26 @@
+
+public class User {
+  public String username;
+  public String comment;
```
In general Java fields should be private. If you need to access these from outside the class you can add a public "getter" method to the class which returns the field value. (eg: public String getUserName() { return this.usrname;} ). 

This touches on a broader topic of encapsulation in object oriented programming. Which is definitely something we'll talk more about as you get into more complex server code, and I'm happy to answer questions about whenever.

TL;DR in this case is that, if these fields are public then a client of this class (eg: DataServlet.java) can directly modify the fields. For a simple data class like User that doesn't matter too much. But in a lot of other situations it can cause problems, so the convention is to always make fields private.

tristan-jh replied:
Another alternative (instead of adding getters) here would be to make User an inner class of DataServlet.java. Encapsulation is not as pressing a concern for inner classes, so in that case it would be ok to leave the fields as 'public'.
Docs on inner classes (also called nested classes): https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html.

tristan-jh commented on portfolio/src/main/webapp/script.js:
```diff
@@ -49,9 +49,23 @@ function randomFood() {
+  fetch('/data').then(response => response.json()).then((users) => {
+    const printCom = document.getElementById('data-container');
+    users.forEach((user) => {
+    printCom.innerHTML = '';
```
resetting printCom.innerHTML here clears comments which were added in earlier iterations of this loop. So this only displays the last comment in 'users'. Is that intended?

adamfweidman replied:
I fixed this in a later CL but you are correct it wasn't

tristan-jh commented on portfolio/src/main/webapp/index.html:
```diff
@@ -85,18 +85,21 @@ <h1>Adam's Personal Portfolio</h1>
         </form>
-      </div>
 
       <button onclick="getData()">See Comments</button>
```
I think you mentioned that you're already thinking about this, but I found it a little confusing that I needed to click submit first, then 'See Comments' to view comments on the page. If you're planning to address that in a later CL then feel free to disregard this comment. otherwise maybe there's a way to get both to happen when you click submit (eg: submit my comment and show others with a single click).

bootsieh commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -20,25 +20,33 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  
+  ArrayList<User> users = new ArrayList<User>();
```
A couple things on this declaration:

1. It's common practice to use type inference in Java now (I think as of Java 7?), so you can write ArrayList<User> users = new ArrayList<>(), and the compiler will fill in the brackets itself.
2. Remember to use the interface List here
3. When making an object field, you want to declare it outside the constructor, but initialize it *inside* the constructor. So you'd have "List<User> users;", and then "users = new ArrayList<User>();" in the constructor.
4. You need to write what classes have access this field. The default for java is 'package-only', which means that every class in the package could call DataServlet.users and potentially modify it. What you want here is 'private', so that the field is only accessible inside the class itself. You'll almost always want your fields set to private.

adamfweidman replied:
I implemented 1, 2, and 4. Later on I don't use a ArrayList to store the information. Declaring the arraylist in the doGet  function resets the array every everytime. So it wouldn't store any of the comments I posted.

tristan-jh reviewed (changes requested):
Looks good! Just a couple java style comments and one question about script.js
